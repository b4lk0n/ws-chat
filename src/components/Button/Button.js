import React from 'react';
import { oneOf, node } from 'prop-types';
import cx from 'classnames';

import { Text } from '../Text';
import styles from './Button.module.css';

export const Button = ({ label, variant, className, ...props }) => (
  <button data-testid={'button'} {...props} className={cx(styles.root, styles[variant], className)}>
    <Text variant={'subtitle'} className={styles.label}>
      {label}
    </Text>
  </button>
);
Button.propTypes = {
  label: node.isRequired,
  variant: oneOf(['outlined', 'primary']),
};
