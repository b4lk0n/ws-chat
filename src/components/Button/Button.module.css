.root {
  --paddingHorizontal: 8px;

  user-select: none;
  border: none;
  border-radius: 4px;
  padding: 6px var(--paddingHorizontal);
  outline: none;
  cursor: pointer;
  background-color: transparent;
  height: 36px;
  color: var(--textPrimary);
  position: relative;
  overflow: hidden;
}

.label {
  position: relative;
}

.root::before {
  display: block;
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;

  transition: background-color 0.15s ease-in-out;
}

.root:not(:disabled):hover::before,
.root:not(:disabled):focus::before {
  background-color: var(--outlineLight);
}

.root:not(:disabled):active::before {
  background-color: var(--outline);
}

.root:disabled {
  opacity: 0.5;
  cursor: default;
}

.outlined,
.primary {
  --paddingHorizontal: 16px;
}

.outlined {
  border: 1px solid var(--outline);
}

.primary {
  background-color: var(--primary);
  color: var(--background);
}
