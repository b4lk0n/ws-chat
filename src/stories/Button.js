import React from 'react';
import { withKnobs, text, select, boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import { Button } from '../components';

export default {
  title: 'Button',
  decorators: [withKnobs],
};

export const Default = () => (
  <Button
    label={text('label', 'Click me')}
    variant={select('variant', {
      'â€”': undefined,
      outlined: 'outlined',
      primary: 'primary',
    })}
    disabled={boolean('disabled')}
    onClick={action('onClick')}
  />
);
