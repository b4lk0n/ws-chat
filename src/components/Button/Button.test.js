import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Button } from './Button';

describe('<Button />', () => {
  it('should renders correctly', () => {
    const clickHandler = jest.fn();
    const { getByTestId } = render(<Button label={'click me'} onClick={clickHandler} />);

    const button = getByTestId('button');
    expect(button).toHaveTextContent('click me');
    expect(clickHandler).not.toBeCalled();
    fireEvent.click(button);
    expect(clickHandler).toBeCalledTimes(1);
    expect(button).toMatchSnapshot();
  });

  ['outlined', 'primary'].forEach(variant => {
    it(`should render ${variant} button`, () => {
      const { getByTestId } = render(<Button label={'click me'} variant={variant} />);
      const button = getByTestId('button');

      expect(button).toHaveTextContent('click me');
      expect(button).toHaveClass(variant);
      expect(button).toMatchSnapshot();
    });
  });

  it('should add additional class if provided', () => {
    const { getByTestId } = render(<Button label={'click me'} className={'bordered'} />);
    const button = getByTestId('button');
    expect(button).toHaveClass('bordered');
    expect(button).toMatchSnapshot();
  });

  it('should not fire events if disabled', () => {
    const clickHandler = jest.fn();
    const { getByTestId } = render(<Button label={'click me'} onClick={clickHandler} disabled />);
    const button = getByTestId('button');
    fireEvent.click(button);
    expect(clickHandler).not.toBeCalled();
    expect(button).toMatchSnapshot();
  });
});
