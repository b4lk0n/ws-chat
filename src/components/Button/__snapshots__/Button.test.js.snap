// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should add additional class if provided 1`] = `
<button
  class="root undefined bordered"
  data-testid="button"
>
  <span
    class="root subtitle label"
    data-testid="text"
  >
    click me
  </span>
</button>
`;

exports[`<Button /> should not fire events if disabled 1`] = `
<button
  class="root undefined"
  data-testid="button"
  disabled=""
>
  <span
    class="root subtitle label"
    data-testid="text"
  >
    click me
  </span>
</button>
`;

exports[`<Button /> should render outlined button 1`] = `
<button
  class="root outlined"
  data-testid="button"
>
  <span
    class="root subtitle label"
    data-testid="text"
  >
    click me
  </span>
</button>
`;

exports[`<Button /> should render primary button 1`] = `
<button
  class="root primary"
  data-testid="button"
>
  <span
    class="root subtitle label"
    data-testid="text"
  >
    click me
  </span>
</button>
`;

exports[`<Button /> should renders correctly 1`] = `
<button
  class="root undefined"
  data-testid="button"
>
  <span
    class="root subtitle label"
    data-testid="text"
  >
    click me
  </span>
</button>
`;
